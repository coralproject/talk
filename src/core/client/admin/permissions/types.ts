import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
