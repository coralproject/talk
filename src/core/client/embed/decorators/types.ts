import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
