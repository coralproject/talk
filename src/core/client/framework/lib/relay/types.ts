import { _RefType } from "react-relay";

export type FragmentKeys<T> = {
  [P in keyof T]: T[P] extends _RefType<any> | null ? P : never
}[keyof T];

export type FragmentKeysNoLocal<T> = Exclude<FragmentKeys<T>, "local">;
