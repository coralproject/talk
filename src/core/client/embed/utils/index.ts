export { default as buildURL } from "./buildURL";
export { default as ensureEndSlash } from "./ensureEndSlash";
export { default as ensureNoEndSlash } from "./ensureNoEndSlash";
export { default as startsWith } from "./startsWith";
export { default as prefixStorage } from "./prefixStorage";
export { default as parseHashQuery } from "./parseHashQuery";
