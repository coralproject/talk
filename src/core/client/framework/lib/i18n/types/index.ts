export { default as FluentShortNumber } from "./FluentShortNumber";
