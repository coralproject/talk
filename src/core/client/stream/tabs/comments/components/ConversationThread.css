.root {
}

.loadMore {
  padding-left: var(--spacing-unit);
  padding-bottom: var(--spacing-unit);
}
