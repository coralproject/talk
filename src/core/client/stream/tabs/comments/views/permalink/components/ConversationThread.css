.root {
}

.loadMore {
  padding-left: var(--spacing-unit);
  padding-bottom: var(--spacing-unit);
}

.showMoreButton {
  text-transform: uppercase;
}
