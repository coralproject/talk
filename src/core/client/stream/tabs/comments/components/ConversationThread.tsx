import cn from "classnames";
import { Localized } from "fluent-react/compat";
import React, { StatelessComponent } from "react";

import { PropTypesOf } from "talk-framework/types";
import { Button, Flex, HorizontalGutter } from "talk-ui/components";

import CommentContainer from "../containers/CommentContainer";
import LocalReplyListContainer from "../containers/LocalReplyListContainer";
import { RootParent } from "./Comment";
import * as styles from "./ConversationThread.css";
import Counter from "./Counter";
import { Circle, Line } from "./Timeline";

export interface ConversationThreadProps {
  className?: string;
  me: PropTypesOf<typeof CommentContainer>["me"] &
    (PropTypesOf<typeof LocalReplyListContainer>["me"] | null);
  asset: PropTypesOf<typeof CommentContainer>["asset"] &
    PropTypesOf<typeof LocalReplyListContainer>["asset"];
  settings: PropTypesOf<typeof CommentContainer>["settings"] &
    PropTypesOf<typeof LocalReplyListContainer>["settings"];
  comment: PropTypesOf<typeof CommentContainer>["comment"];
  disableLoadMore: boolean;
  loadMore: () => void;
  remaining: number;
  parents: Array<
    { id: string } & PropTypesOf<typeof CommentContainer>["comment"] &
      PropTypesOf<typeof LocalReplyListContainer>["comment"]
  >;
  rootParent: {
    id: string;
    createdAt: string;
    username: string | null;
  } | null;
}

const ConversationThread: StatelessComponent<
  ConversationThreadProps
> = props => {
  if (props.remaining === 0 && props.parents.length === 0) {
    return (
      <div className={cn(props.className, styles.root)}>
        <CommentContainer
          comment={props.comment}
          asset={props.asset}
          settings={props.settings}
          me={props.me}
          highlight
        />
      </div>
    );
  }
  return (
    <div className={cn(props.className, styles.root)}>
      <HorizontalGutter container={<Line dotted />}>
        {props.rootParent && (
          <Circle>
            <RootParent
              id={props.rootParent.id}
              username={props.rootParent.username}
              createdAt={props.rootParent.createdAt}
            />
          </Circle>
        )}
        {props.remaining > 0 && (
          <Circle hollow className={styles.loadMore}>
            <Flex alignItems="center" itemGutter="half">
              <Localized
                id="comments-conversationThread-showHiddenComments"
                $count={props.remaining}
              >
                <Button
                  onClick={props.loadMore}
                  disabled={props.disableLoadMore}
                  id="comments-conversationThread-showHiddenComments"
                  variant="underlined"
                >
                  Show hidden comments
                </Button>
              </Localized>
              {props.remaining > 1 && <Counter>{props.remaining}</Counter>}
            </Flex>
          </Circle>
        )}
      </HorizontalGutter>
      <HorizontalGutter container={Line}>
        {props.parents.map((parent, i) => (
          <Circle key={parent.id} hollow={!!props.remaining || i > 0}>
            <CommentContainer
              comment={parent}
              asset={props.asset}
              me={props.me}
              settings={props.settings}
              localReply
            />
            {props.me && (
              <LocalReplyListContainer
                asset={props.asset}
                me={props.me}
                settings={props.settings}
                comment={parent}
                indentLevel={1}
              />
            )}
          </Circle>
        ))}
        <Circle end>
          <CommentContainer
            comment={props.comment}
            asset={props.asset}
            settings={props.settings}
            me={props.me}
            highlight
          />
        </Circle>
      </HorizontalGutter>
    </div>
  );
};

export default ConversationThread;
