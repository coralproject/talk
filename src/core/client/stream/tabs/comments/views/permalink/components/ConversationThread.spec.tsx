import { noop } from "lodash";
import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { pureMerge } from "talk-common/utils";
import { removeFragmentRefs } from "talk-framework/testHelpers";
import { DeepPartial, PropTypesOf } from "talk-framework/types";

import ConversationThread from "./ConversationThread";

const ConversationThreadN = removeFragmentRefs(ConversationThread);

type Props = PropTypesOf<typeof ConversationThreadN>;

function createDefaultProps(add: DeepPartial<Props> = {}): Props {
  return pureMerge(
    {
      className: "root",
      viewer: {},
      story: {},
      settings: {},
      comment: {},
      disableLoadMore: false,
      loadMore: noop,
      remaining: 2,
      parents: [],
      rootParent: {
        id: "root-parent",
        createdAt: "1995-12-17T03:24:00.000Z",
        username: "parentAuthor",
        tags: [],
      },
    },
    add
  );
}

describe("with 2 remaining parent comments", () => {
  it("renders correctly", () => {
    const props = createDefaultProps();
    const renderer = createRenderer();
    renderer.render(<ConversationThreadN {...props} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it("renders staff badge", () => {
    const props = createDefaultProps({
      rootParent: {
        tags: ["Staff"],
      },
    });
    const renderer = createRenderer();
    renderer.render(<ConversationThreadN {...props} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it("renders with disabled load more", () => {
    const props = createDefaultProps({
      disableLoadMore: true,
    });
    const renderer = createRenderer();
    renderer.render(<ConversationThreadN {...props} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});

it("renders with no parent comments", () => {
  const props = createDefaultProps({
    remaining: 0,
    rootParent: null,
  });
  const renderer = createRenderer();
  renderer.render(<ConversationThreadN {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
