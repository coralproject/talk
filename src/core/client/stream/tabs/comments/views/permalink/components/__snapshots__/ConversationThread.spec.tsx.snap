// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with no parent comments 1`] = `
<div
  className="root ConversationThread-root"
  data-testid="comments-permalinkView-conversationThread"
>
  <withContext(createMutationContainer(withContext(createMutationContainer(Relay(CommentContainer)))))
    comment={Object {}}
    highlight={true}
    me={Object {}}
    settings={Object {}}
    story={Object {}}
  />
</div>
`;

exports[`with 2 remaining parent comments renders correctly 1`] = `
<div
  className="root ConversationThread-root"
  data-testid="comments-permalinkView-conversationThread"
>
  <ForwardRef(forwardRef)
    container={
      <Line
        dotted={true}
      />
    }
  >
    <Circle>
      <RootParent
        createdAt="1995-12-17T03:24:00.000Z"
        id="root-parent"
        username="parentAuthor"
      />
    </Circle>
    <Circle
      className="ConversationThread-loadMore"
      hollow={true}
    >
      <ForwardRef(forwardRef)
        alignItems="center"
        itemGutter="half"
      >
        <Localized
          $count={2}
          id="comments-conversationThread-showMoreOfThisConversation"
        >
          <ForwardRef(forwardRef)
            className="ConversationThread-showMoreButton"
            disabled={false}
            onClick={[Function]}
            variant="underlined"
          >
            Show more of this conversation
          </ForwardRef(forwardRef)>
        </Localized>
        <withPropsOnChange(Counter)>
          2
        </withPropsOnChange(Counter)>
      </ForwardRef(forwardRef)>
    </Circle>
  </ForwardRef(forwardRef)>
  <ForwardRef(forwardRef)
    container={[Function]}
  >
    <Circle
      end={true}
    >
      <withContext(createMutationContainer(withContext(createMutationContainer(Relay(CommentContainer)))))
        comment={Object {}}
        highlight={true}
        me={Object {}}
        settings={Object {}}
        story={Object {}}
      />
    </Circle>
  </ForwardRef(forwardRef)>
</div>
`;

exports[`with 2 remaining parent comments renders with disabled load more 1`] = `
<div
  className="root ConversationThread-root"
  data-testid="comments-permalinkView-conversationThread"
>
  <ForwardRef(forwardRef)
    container={
      <Line
        dotted={true}
      />
    }
  >
    <Circle>
      <RootParent
        createdAt="1995-12-17T03:24:00.000Z"
        id="root-parent"
        username="parentAuthor"
      />
    </Circle>
    <Circle
      className="ConversationThread-loadMore"
      hollow={true}
    >
      <ForwardRef(forwardRef)
        alignItems="center"
        itemGutter="half"
      >
        <Localized
          $count={2}
          id="comments-conversationThread-showMoreOfThisConversation"
        >
          <ForwardRef(forwardRef)
            className="ConversationThread-showMoreButton"
            disabled={true}
            onClick={[Function]}
            variant="underlined"
          >
            Show more of this conversation
          </ForwardRef(forwardRef)>
        </Localized>
        <withPropsOnChange(Counter)>
          2
        </withPropsOnChange(Counter)>
      </ForwardRef(forwardRef)>
    </Circle>
  </ForwardRef(forwardRef)>
  <ForwardRef(forwardRef)
    container={[Function]}
  >
    <Circle
      end={true}
    >
      <withContext(createMutationContainer(withContext(createMutationContainer(Relay(CommentContainer)))))
        comment={Object {}}
        highlight={true}
        me={Object {}}
        settings={Object {}}
        story={Object {}}
      />
    </Circle>
  </ForwardRef(forwardRef)>
</div>
`;
